python/ray/data/_internal/execution/operators/map_operator.py

```diff
@@ -530,17 +530,17 @@ def _map_task(
     """
     DataContext._set_current(data_context)
     ctx.kwargs.update(kwargs)
-    stats = BlockExecStats.builder()
     map_transformer.set_target_max_block_size(ctx.target_max_block_size)
-    for b_out in map_transformer.apply_transform(iter(blocks), ctx):
-        # TODO(Clark): Add input file propagation from input blocks.
-        m_out = BlockAccessor.for_block(b_out).get_metadata()
-        m_out.exec_stats = stats.build()
-        m_out.exec_stats.udf_time_s = map_transformer.udf_time()
-        m_out.exec_stats.task_idx = ctx.task_idx
-        yield b_out
-        yield m_out
-        stats = BlockExecStats.builder()
+    with _BlockExecStatsBuilder(data_context.memory_poll_interval_s) as stats:
+        for b_out in map_transformer.apply_transform(iter(blocks), ctx):
+            # TODO(Clark): Add input file propagation from input blocks.
+            m_out = BlockAccessor.for_block(b_out).get_metadata()
+            m_out.exec_stats = stats.build()
+            m_out.exec_stats.udf_time_s = map_transformer.udf_time()
+            m_out.exec_stats.task_idx = ctx.task_idx
+            yield b_out
+            yield m_out
+            stats.reset()
```

Comment:
@bveeramani can we actually abstract memory tracking into its own component and pass it into BlockStatsBuilder. 

I think it'll be better to keep StatsBuilder itself simpler 

  - Their lifecycles are quite different: we want to reset the stats for every block, but not the memory tracker
  - We keep their domain disjoint (though we'll pass memory tracker to builder to get the data from)